crates/sui-package-resolver/src/store.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+use std::path::Path;
+use std::sync::Arc;
+
+use crate::error::Error;
+use crate::Result;
+use crate::{Module, Package};
+use async_trait::async_trait;
+use diesel::{ExpressionMethods, OptionalExtension, QueryDsl, RunQueryDsl};
+use move_binary_format::{access::ModuleAccess, errors::Location, CompiledModule};
+use move_core_types::account_address::AccountAddress;
+use sui_indexer::{indexer_reader::IndexerReader, schema_v2::objects};
+use sui_rest_api::Client;
+use sui_types::base_types::ObjectID;
+use sui_types::{base_types::SequenceNumber, move_package::TypeOrigin, object::Object};
+use typed_store::rocks::{DBMap, MetricConf};
+use typed_store::traits::TableSummary;
+use typed_store::traits::TypedStoreDebug;
+use typed_store::Map;
+use typed_store_derive::DBMapUtils;
+
+/// Interface to abstract over access to a store of live packages.  Used to override the default
+/// store during testing.
+#[async_trait]
+pub trait PackageStore {
```

Comment:
@amnn I think we may want to discuss retooling this trait and the coupled Cache. As it stands today there are a few issues I see with it.

- The `version` function is only used to check for newer versions of the system packages since they are the only packages that can increase in version number. This event of them increasing in version number is a very, very rare occurrence and yet we're currently paying a rather large penalty, a single round-trip, for every single occurrence of a system package type when resolving a type. Given the framework types are far more likely to show up than any others this seems like a bad trade-off for a cache. I'd propose we remove this mechanism entirely and instead provide a method on  the cache for forcefully evicting packages from the cache so that policies for how/when we fallback to the refresh system packages can be run outside of the critical path of doing a type resolution.
- @sadhansood The addition of the `update` function seems out of place for what i perceive to be the intent behind this trait which is that the backing package store is to be used as a read-only store and not something that you can update. Instead I think that this `update` function seems like a prime candidate to not be apart of this trait but instead a method on the `PackageCache` itself used to update the local cache state, preventing the need to reach out to some remote store for the update. If you do indeed want to have a way to update the backing package store, in the case of having a rocksdb local store, i'd hold onto a handle to the store and update it directly there.

Reply:
Looks like we raced on reviews!

> The version function is only used to check for newer versions of the system packages since they are the only packages that can increase in version number.

> I'd propose we remove this mechanism entirely and instead provide a method on the cache for forcefully evicting packages from the cache so that policies for how/when we fallback to the refresh system packages can be run outside of the critical path of doing a type resolution.

A system to forcefully evict packages works well in the data pipeline cases where data is being pushed through the pipeline, but it does not work so well for the GraphQL service which is pulling data, so has no idea when to evict.  Any issues with the plan we discussed in Denver which was to put a type layout cache in front of the package cache.?  Framework types, which show up often, will remain in this cache.  Therefore we won't hit the slow path with the version check in it often, or at all in practice, unless somebody queries a layout involving a brand new framework type.

Also open to suggestions that would allow the pipelines to evict on push, but the GraphQL service to re-query on pull.  In another comment on this PR, I suggested fully encapsulating the storage interface (i.e. no `LruCache` by default) maybe that gets us what we want along this vein?

> The addition of the update function seems out of place for what i perceive to be the intent behind this trait which is that the backing package store is to be used as a read-only store and not something that you can update.

I tend to agree here as well -- this was another thing that made me want to change how the store interface worked here (although, for what it's worth, I was also planning on doing something similar to this in the GraphQL service, for the case where a package is loaded as an object first, and then queried for its structs and functions -- it's useful in that case to be able to inject a package into the "store", but in the GraphQL case the main benefit is to update the `LruCache`, not touch the backing store).

Reply:
> but it does not work so well for the GraphQL service which is pulling data, so has no idea when to evict

We had a discussion a bit ago about how subscriptions would be handled in the GraphQL service, specifically that one approach would be for the GraphQL service to poll the DB for updates. If we go that route then it seems like we could also leverage that same mechanism for when to evict system packages

Reply:
The difference for me is that the push-pull mechanism for subscriptions means that for downstream clients, it is the "leading edge" that causes them to query further things, and all we need to do is make sure that before we issue the subscription update, everything else in the store is consistent (and polling is fine/good for this).  If we use the same mechanism for updating the underlying store (in this cache the type layout "store"), a query could fail because we didn't poll for the system package update fast enough, although everything else is otherwise consistent.